runtime/vm/xa64/UpcallThunkGen.cpp

```diff
@@ -35,32 +35,1219 @@ extern "C" {
+					S8_mRSP_DISP32m_SREG(thunkCursor, frameOffsetCursor, gprParmRegs[gprRegParmCount])
+					gprRegParmCount++;
+				} else {
+					// Parm must be filled from frame and spilled to argList.
+					// Use rax as the intermediary register since it is volatile
+					L8_TREG_mRSP_DISP32m(thunkCursor, rax, frameSize + preservedRegisterAreaSize + 8 + memParmCursor)
+					S8_mRSP_DISP32m_SREG(thunkCursor, frameOffsetCursor, rax)
+					memParmCursor += STACK_SLOT_SIZE;
+				}
+
+				frameOffsetCursor += STACK_SLOT_SIZE;
+				break;
+			}
+			case J9_FFI_UPCALL_SIG_TYPE_FLOAT:  // Fall through
+			case J9_FFI_UPCALL_SIG_TYPE_DOUBLE:
+			{
+				bool isFloat = (sigArray[i].type == J9_FFI_UPCALL_SIG_TYPE_FLOAT);
+				if (fprRegParmCount < MAX_FPRS_PASSED_IN_REGS) {
+					// Parm must be spilled from parm register to argList
+					if (isFloat) {
+						MOVSS_mRSP_DISP32m_SREG(thunkCursor, frameOffsetCursor, fprParmRegs[fprRegParmCount])
+					} else {
+						MOVSD_mRSP_DISP32m_SREG(thunkCursor, frameOffsetCursor, fprParmRegs[fprRegParmCount])
+					}
+
+					fprRegParmCount++;
+				} else {
+					// Parm must be filled from frame and spilled to argList.
+					// Use xmm0 as the intermediary register since it is volatile and it
+					// must have been processed as the first parameter already.
+					if (isFloat) {
+						MOVSS_TREG_mRSP_DISP32m(thunkCursor, xmm0, frameSize + preservedRegisterAreaSize + 8 + memParmCursor)
+						MOVSS_mRSP_DISP32m_SREG(thunkCursor, frameOffsetCursor, xmm0)
+					} else {
+						MOVSD_TREG_mRSP_DISP32m(thunkCursor, xmm0, frameSize + preservedRegisterAreaSize + 8 + memParmCursor)
+						MOVSD_mRSP_DISP32m_SREG(thunkCursor, frameOffsetCursor, xmm0)
+					}
+
+					memParmCursor += STACK_SLOT_SIZE;
+				}
+
+				frameOffsetCursor += STACK_SLOT_SIZE;
+				break;
+			}
+			default:
+			{
+				// Handle structs passed in registers.  Structs passed in memory will be handled
+				// after all other parameters are processed to avoid the complication of preserving
+				// registers implicitly required for REP MOVSB.
+				X64StructPassingMechanism mechanism = analyzeStructParm(gprRegParmCount, fprRegParmCount, sigArray[i]);
+				switch (mechanism) {
+					case PASS_STRUCT_IN_MEMORY:
+					{
+						// Record the source and destination offsets and the number of bytes to copy
+						structParmInMemory[numStructsPassedInMemoryCursor].memParmCursor = memParmCursor;
+						structParmInMemory[numStructsPassedInMemoryCursor].frameOffsetCursor = frameOffsetCursor;
+						structParmInMemory[numStructsPassedInMemoryCursor].sizeofStruct = sigArray[i].sizeInByte;
+
+						I_32 roundedStructSize = ROUND_UP_TO_SLOT_MULTIPLE(sigArray[i].sizeInByte);
+						memParmCursor += roundedStructSize;
+						frameOffsetCursor += roundedStructSize;
+						numStructsPassedInMemoryCursor += 1;
+						break;
+					}
+					case PASS_STRUCT_IN_ONE_FPR:
+					{
+						MOVSD_mRSP_DISP32m_SREG(thunkCursor, frameOffsetCursor, fprParmRegs[fprRegParmCount])
```

Comment:
do you need to handle single float member vs.  two float member and single double member differently here?

Reply:
the point is that transferring 64bit out of FPR for a single-precision float is wrong for the memory image, isn't it?

Reply:
I tinkered with some assembly and the debugger today to observe what happens in these cases and it does what I expect.  Are you worried that when the 64-bits (bits 0-63) are stored and then read as a 32-bit float (bits 0-31) that the wrong half (bits 32-63) are read instead?  If not, can you describe the case you're concerned with?

Reply:
yes, related.  it all depends on how and where in FPR a single-precision float is represented by movss.  see the two examples below whether you need to handle them differently:
```
struct threeFloats {
   float a[3];
}

struct fourFloats {
   float a[4];
}
```
C establishes threeFloats and fourFloats (if they are passed in FPRs) differently movsd + movss  vs.  movsd + movsd). Your current code treats them the same.

Reply:
on Power, definitely they need to be treated differently for sure.

Reply:
I verified with a C program with custom assembly to mimic how the thunk marshalls these struct arguments (including the threeFloats and fourFloats examples) and the thunk behaves as expected when using MOVSD instructions.  I don't think there is an issue here.